### PR TITLE
[CI] Cancel workflows for stale commits automatically

### PR DESCRIPTION
**Pull request by kmichaelk**

### CODE REVIEW

**Comment by kmichaelk** [comment]
@aobolensk

**Comment by aobolensk** [comment]
To be in equal position for everyone, let us merge this in 10 days (last day when you can get points for the 1st task). November, 13.

File: .github/workflows/perf-statistic.yml [hunk 1]
```diff
@@ -6,6 +6,10 @@ on:
   schedule:
     - cron: '0 12 * * *'
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
```
Comment (aobolensk) [inline .github/workflows/perf-statistic.yml:12]:
```suggestion
```
Please revert this change for perf stats, it is not correct to have it there

**Review by aobolensk — CHANGES_REQUESTED** [review]

**Review by allnes — APPROVED** [review]
Agree with @aobolensk, we will wait 13th November.

**Review by aobolensk — APPROVED** [review]

**Comment by aobolensk** [comment]
@kmichaelk LGTM, thanks. If you don't mind, please consider adding your infrastructure related changes to the main repository: https://github.com/learning-process/parallel_programming_course . This will be useful for the next term when repo will be re-created and for next year students.

Thanks in advance!

**Comment by kmichaelk** [comment]
Of course.

I'd also suggest to get rid of extensive use of `std::shared_ptr` - e.g., in the case with `TaskData`, its lifetime is not tied to the lifetime of the memory it points to and I don't think it is needed at all.

**Pull request merged**